src/main/cpp/io/ArmIO.cpp

```diff
@@ -32,23 +32,83 @@ bool ArmIO::ProcessIO() {
-    hw_interface_->arm_lift_motor->Set(sw_interface_->arm_lift_power);
-    hw_interface_->arm_up_motor->Set(sw_interface_->arm_up_power);
-    hw_interface_->arm_extend_motor->Set(sw_interface_->arm_extend_power);
+   
+    sw_interface_->arm_encoder = hw_interface_->arm_lift_encoder->GetPosition();
+    sw_interface_->arm_extend_encoder = hw_interface_->arm_extend_encoder->GetPosition();
+    sw_interface_->arm_duty_cycle_encoder = hw_interface_->arm_duty_cycle_encoder->GetAbsolutePosition() * 360  -  330; // offset of 330
+
+    if (sw_interface_->arm_duty_cycle_encoder < 0) {
+        sw_interface_->arm_duty_cycle_encoder += 360;
+        sw_interface_->arm_duty_cycle_encoder *= -1;
+    }
+    
+    TeamOKC::Clamp(-0.2, 0.2, &sw_interface_->arm_lift_power);
+    TeamOKC::Clamp(-0.2, 0.2, &sw_interface_->arm_extend_power);
+
+    // if the absolute encoder is >110 degrees
+    if (sw_interface_->arm_duty_cycle_encoder >= 110) {
+        // and we're trying to go farther positive
+        if (sw_interface_->arm_lift_power > 0) {
+            // stop it
+            hw_interface_->arm_lift_motor->Set(0);
+            hw_interface_->arm_up_motor->Set(0);
+        } else {
+            // otherwise, it's cool
+            hw_interface_->arm_lift_motor->Set(sw_interface_->arm_lift_power);
+            hw_interface_->arm_up_motor->Set(-sw_interface_->arm_lift_power);
+        }
+    } else if (sw_interface_->arm_duty_cycle_encoder <= -110) {
+        // and we're trying to go farther negative
+        if (sw_interface_->arm_lift_power < 0) {
+            // stop it
+            hw_interface_->arm_lift_motor->Set(0);
+            hw_interface_->arm_up_motor->Set(0);
+        } else {
+            // otherwise, it's cool
+            hw_interface_->arm_lift_motor->Set(sw_interface_->arm_lift_power);
+            hw_interface_->arm_up_motor->Set(-sw_interface_->arm_lift_power);
+        }
+    } else { // if neither limit is reached
+        hw_interface_->arm_lift_motor->Set(sw_interface_->arm_lift_power);
+        hw_interface_->arm_up_motor->Set(-sw_interface_->arm_lift_power);
+    }
+
+    sw_interface_->extend_limit_switch = !hw_interface_->extend_limit_switch->Get(); // limit switch is inverse logic
+
+    // if the IR sensor is getting a read
+    if (sw_interface_->extend_limit_switch) {
+        // reset arm encoder
+        hw_interface_->arm_extend_encoder->SetPosition(0.0);
+
+        // if the motor power is positive
+        if (sw_interface_->arm_extend_power > 0) {
+            // let it pass, so the arm can move outwards
+            hw_interface_->arm_extend_motor->Set(sw_interface_->arm_extend_power);
+        } else {
+            // otherwise, limit it to 0 and don't let it move
+            hw_interface_->arm_extend_motor->Set(0);
+        }
+    } else { // if the limit switch isn't pressed
+        // if we're instead at the farthest point
+        if (sw_interface_->arm_extend_encoder > 100) {
```

Comment:
Should we parameterize this 100 in the TOML file? If not, then at the very least I think these magic numbers in this file need to be `#define`d in ArmIO.h so they have names associated with them

Reply:
probably? during testing it was actually the perfect number somehow, but yeah it should be somewhere else